File: tools_parity_check.js
/* JS 引擎 vs Python 版的對照測試（Node 執行）
 *
 * 兩套實作最怕的是「漂移」：改了一邊忘了另一邊，公開站的分數就跟本機工具對不起來。
 * 這支把 Python 算出來的基準（tools_parity_check.py 產生的 parity_expected.json）
 * 拿來跟 JS 引擎的輸出逐項比對，超過容差就以非零結束碼失敗。
 */
const fs = require('fs');
const path = require('path');
const vm = require('vm');
/* 用 vm.runInThisContext 而不是 eval：engine.js 開頭有 "use strict"，
   在嚴格模式的 eval 裡宣告的函式不會外洩到全域，載進來會等於什麼都沒有。
   以 Script 形式執行則頂層宣告仍會建立全域繫結。 */
const runGlobal = (code, name) => vm.runInThisContext(code, { filename: name });
const ROOT = path.dirname(__filename);
const SITE = path.join(ROOT, 'site');

/* ── Worker 環境模擬 ── */
const msgs = [];
global.self = {
  postMessage: (m) => msgs.push(m),
  set onmessage(fn) { global.__onmessage = fn; },
  get onmessage() { return global.__onmessage; },
};
global.importScripts = (f) => runGlobal(fs.readFileSync(path.join(SITE, f), 'utf8'), f);
global.fetch = async (u) => {
  if (u.startsWith('http')) return { ok: false, status: 0 };   // 高程 API：離線測試一律略過
  const p = path.join(SITE, u);
  if (!fs.existsSync(p)) return { ok: false, status: 404 };
  const buf = fs.readFileSync(p);
  return {
    ok: true, status: 200,
    async json() { return JSON.parse(buf.toString('utf8')); },
    async arrayBuffer() { return buf.buffer.slice(buf.byteOffset, buf.byteOffset + buf.byteLength); },
  };
};

runGlobal(fs.readFileSync(path.join(SITE, 'worker.js'), 'utf8'), 'worker.js');

const EXP = JSON.parse(fs.readFileSync(path.join(ROOT, 'out', 'parity_expected.json'), 'utf8'));
let fails = 0;
function cmp(label, got, want, tol, unit) {
  const d = Math.abs(got - want);
  const ok = d <= tol;
  if (!ok) fails++;
  console.log(`  ${ok ? 'OK  ' : 'FAIL'} ${label.padEnd(22)} JS ${String(got).padStart(9)}` +
    `  Python ${String(want).padStart(9)}  差 ${d.toFixed(3)}${unit || ''}（容差 ${tol}）`);
}

(async () => {
  /* ── A. 描圖對照 ── */
  console.log('\nA. 描圖（同一張手繪圖）');
  const g = JSON.parse(fs.readFileSync(path.join(ROOT, 'out', 'parity_gray.json'), 'utf8'));
  const gray = Float32Array.from(g.gray);
  const tr = traceImage(gray, g.w, g.h, 6, 64);
  cmp('寬高比', tr.info['寬高比'], EXP.trace.aspect, 0.06);
  cmp('簡化後點數', tr.info['簡化後點數'], EXP.trace.points, 12, ' 點');
  // 形狀相似度：把兩條輪廓各取樣 200 點，比對彼此的平均最近距離（正規化座標）
  const A = resample(tr.pts, 200), B = resample(EXP.trace.pts, 200);
  let s = 0;
  for (const p of A) s += polyDist(p[0], p[1], B);
  cmp('輪廓平均偏離', Math.round(s / A.length * 10000) / 10000, 0, 0.02);

  /* ── B. 規劃對照（餵同一個圖形，隔離描圖差異）── */
  console.log('\nB. 規劃（南區 × Python 描出的圖形 × 標準級）');
  const t0 = Date.now();
  const done = new Promise((res) => {
    const iv = setInterval(() => {
      const d = msgs.find((m) => m.type === 'done' || m.type === 'error');
      if (d) { clearInterval(iv); res(d); }
    }, 100);
  });
  self.onmessage({
    data: {
      cmd: 'plan', districts: ['south'], level: 'standard',
      shape: EXP.plan.shape, shapeName: '手繪蹲坐貓'
    }
  });
  const d = await done;
  if (d.type === 'error') { console.log('  FAIL 規劃失敗：' + d.message); process.exit(1); }
  const r = d.results[0];
  console.log(`  （JS 耗時 ${((Date.now() - t0) / 1000).toFixed(1)} 秒，Python ${EXP.plan.seconds} 秒）`);
  cmp('里程 km（同級距即可）', r.km, EXP.plan.km, 1.2, ' km');
  cmp('綜合評分（品質相當即可）', r.score.total, EXP.plan.total, 4, ' 分');
  for (const k of ['shape', 'safety', 'distance', 'district', 'logistics']) {
    cmp('構面 ' + k, r.score.parts[k], EXP.plan.parts[k], 8, ' 分');
  }
  cmp('Cue Sheet 段數', r.cues.length, EXP.plan.cues, 20, ' 段');
  cmp('安全提醒則數', r.notes.length, EXP.plan.notes, 5, ' 則');


  /* ── C. 固定佈局：這一段必須嚴格對上 ──
     搜尋是啟發式的，兩邊在平手處挑到不同路線是正常的（上面用寬容差）。
     但同一組（中心、尺寸、旋轉）建出來的路線，評分必須一模一樣，
     不然就是評分本身算錯了。 */
  console.log('');
  console.log('C. 固定佈局（隔離搜尋差異，這段要嚴格一致）');
  const F = EXP.fixed;
  const N = LOADED['south'];
  const fx = build(N, EXP.plan.shape, F.center[0], F.center[1], F.size, F.rot);
  if (!fx) { console.log('  FAIL 固定佈局建不出路線'); process.exit(1); }
  const fsc = scoreRoute(N, fx, 'standard', null);
  /* 容差不是 0 的原因：路網匯出時座標量化到百萬分之一度（約 11 公分），
     而 OSM 上常有相距幾公分的重合節點（兩條路的交會點）。量化後排序可能翻面，
     吸附就會挑到另一個節點，路線因此差幾十到一百公尺。這是資料精度的物理下限，
     不是演算法不一致——總分只差 0.2 分就是證據。 */
  cmp('里程 km', Math.round(fx.length) / 1000, F.km, 0.25, ' km');
  cmp('綜合評分', fsc.total, F.total, 0.6, ' 分');
  for (const k2 of ['shape', 'safety', 'district', 'logistics']) {
    cmp('構面 ' + k2, fsc.parts[k2], F.parts[k2], 1.0, ' 分');
  }
  // distance 構面直接由里程換算，里程的量化誤差會等比放大到這裡
  cmp('構面 distance', fsc.parts.distance, F.parts.distance, 3.0, ' 分');
  cmp('Cue Sheet 段數', cueSheet(N, fx).length, F.cues, 2, ' 段');
  cmp('安全提醒則數', safetyNotes(N, fx).length, F.notes, 2, ' 則');

  console.log(fails ? `\n✗ ${fails} 項超出容差` : '\n✓ 全部在容差內，兩套實作一致');
  process.exit(fails ? 1 : 0);
})();

File: tools_parity_check.py
# -*- coding: utf-8 -*-
"""產生 JS↔Python 對照測試的基準值。

跑法：
    python tools_parity_check.py      產生 out/parity_expected.json 與 out/parity_gray.json
    node tools_parity_check.js        用那份基準檢查 JS 引擎

任何一邊改了演算法都要重跑這兩支。兩套實作漂移是這個架構最大的風險——
公開站算 82 分、本機工具算 79 分，沒有人會知道哪個是對的。
"""
import io
import json
import os
import sys
import time

sys.stdout.reconfigure(encoding="utf-8")
sys.path.insert(0, os.path.dirname(os.path.abspath(__file__)))

import numpy as np                                  # noqa: E402
from PIL import Image                               # noqa: E402

from catrun import shapes                           # noqa: E402
from catrun.planner import plan                     # noqa: E402
from catrun.trace import trace                      # noqa: E402

ROOT = os.path.dirname(os.path.abspath(__file__))
IMG = r"C:/Users/User/Desktop/S__19505160.jpg"
OUT = os.path.join(ROOT, "out")


def main():
    os.makedirs(OUT, exist_ok=True)
    print("1. 描圖基準")
    pts, info = trace(IMG, close_px=6, target_pts=64)
    print("   點數 %d、寬高比 %.3f" % (len(pts), info["寬高比"]))

    # 把灰階影像原樣交給 JS，兩邊才是吃同一份輸入。
    # 灰階值就是 0~255，用 base64 存 Uint8 即可——存成浮點數 JSON 會膨脹到好幾 MB，
    # 大到讓人想刪掉它，結果測試就跑不動了（真的發生過）。
    import base64
    a = np.asarray(Image.open(IMG).convert("L"), dtype=np.uint8)
    with io.open(os.path.join(OUT, "parity_gray.json"), "w", encoding="utf-8") as f:
        json.dump({"w": int(a.shape[1]), "h": int(a.shape[0]),
                   "b64": base64.b64encode(a.tobytes()).decode()}, f)
    print("   灰階影像 %dx%d 已輸出（%.0f KB）"
          % (a.shape[1], a.shape[0],
             os.path.getsize(os.path.join(OUT, "parity_gray.json")) / 1024))

    print("2. 規劃基準（南區 × cat_hand × 標準級）")
    t0 = time.time()
    r = plan("south", "cat_hand", "standard", log=lambda *a2: None, save=False)
    secs = round(time.time() - t0, 1)
    sc = r["score"]
    print("   %.2f km、%.1f 分、%.1f 秒" % (r["meta"]["km"], sc["total"], secs))

    # 對照測試要能離線重跑，所以基準另外算一份「不計高程」的分數：
    # JS 端在離線測試時拿不到 Open-Meteo，兩邊都用 climb=None 才比得準。
    from catrun import fitter, scoring
    sc0 = scoring.score(r["net"], r["route"], "standard", "south", None)

    # ── 固定佈局的基準 ──
    # 搜尋是啟發式的，兩套實作在平手處會挑到不同路線，那是正常的。
    # 但「同一條路線的評分」必須一模一樣，否則就是真的算錯了。
    # 所以另外用一組寫死的（中心、尺寸、旋轉）建路線，這一項要嚴格對上。
    print("3. 固定佈局基準（隔離搜尋的隨機性）")
    net = r["net"]
    base = shapes.normalised("cat_hand")
    s0, w0, n0, e0 = net.cfg["bbox"]
    center = (s0 + (n0 - s0) * 0.5, w0 + (e0 - w0) * 0.5)
    size, rot = 1650.0, -8.0
    fixed = fitter.build(net, base, center, size, rot)
    if fixed is None:
        raise SystemExit("固定佈局建不出路線，換一組參數")
    fsc = scoring.score(net, fixed, "standard", "south", None)
    from catrun import export as _ex
    fcues = _ex.cue_sheet(fixed)
    fnotes = _ex.safety_notes(net, fixed)
    print("   %.3f km、%.1f 分、cue %d 段" % (fixed.km, fsc["total"], len(fcues)))
    exp = {
        "trace": {"points": info["簡化後點數"], "aspect": info["寬高比"], "pts": pts},
        "plan": {
            "shape": shapes.normalised("cat_hand"),
            "km": r["meta"]["km"], "total": sc0["total"], "parts": sc0["parts"],
            "cues": len(r["cues"]), "notes": len(r["notes"]), "seconds": secs,
            "total_with_climb": sc["total"], "climb": r["meta"]["climb_m"],
        },
        "fixed": {
            "center": [center[0], center[1]], "size": size, "rot": rot,
            "km": round(fixed.km, 3), "total": fsc["total"], "parts": fsc["parts"],
            "cues": len(fcues), "notes": len(fnotes),
            "detail": fsc["detail"],
        },
    }
    p = os.path.join(OUT, "parity_expected.json")
    with io.open(p, "w", encoding="utf-8") as f:
        json.dump(exp, f, ensure_ascii=False)
    print("   基準寫出：%s" % p)
    print("\n接著跑：node tools_parity_check.js")


if __name__ == "__main__":
    main()
